### ISSUE TITLE
saved models fail to restore when using Voigt or other models that use complex constraints

### ISSUE THREAD

**Issue by newville** [lmfit/lmfit-py#553]
Saving a model result with a Voigt Function works, but the saved model cannot be restored.

The problem can be reproduced with

``` python
import numpy as np
import matplotlib.pyplot as plt
from lmfit.model import save_modelresult, load_modelresult
from lmfit.models import  VoigtModel
from lmfit.lineshapes import gaussian, lorentzian

x = np.linspace(0, 20, 501)
y = gaussian(x, 1.1, 8.5, 2) + lorentzian(x, 1.7, 8.5, 1.5)
y = y + np.random.normal(size=len(x), scale=0.025)

model = VoigtModel()
pars = model.guess(y, x=x)
result = model.fit(y, pars, x=x)
save_modelresult(result, 'voigt_modelresult.sav')

r2 = load_modelresult('voigt_modelresult.sav')
```

which will give

```
TypeError
   <_ast.Module object at 0x1a263afb00>
                                                               ^^^
'{'__class__': 'Callable', '__name__': 'wofz', 'pyversion': '3.7', 'value': None, 'importer': 'scipy.special'}' is not callable!!
Traceback (most recent call last):
  File "d1.py", line 31, in <module>
    r2 = load_modelresult('voigt_modelresult.sav')
  File "/Users/Newville/Codes/lmfit-py/lmfit/model.py", line 1261, in load_modelresult
    mresult = modres.load(fh, funcdefs=funcdefs)
  File "/Users/Newville/Codes/lmfit-py/lmfit/model.py", line 1720, in load
    return self.loads(fp.read(), funcdefs=funcdefs, **kws)
  File "/Users/Newville/Codes/lmfit-py/lmfit/model.py", line 1676, in loads
    self.params.__setstate__(state)
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 205, in __setstate__
    self.add_many(*state['params'])
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 393, in add_many
    self.__setitem__(para.name, para)
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 131, in __setitem__
    self._asteval.symtable[key] = par.value
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 820, in value
    return self._getval()
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 802, in _getval
    check_ast_errors(self._expr_eval)
  File "/Users/Newville/Codes/lmfit-py/lmfit/parameter.py", line 24, in check_ast_errors
    expr_eval.raise_exception(None)
  File "/Users/Newville/anaconda3/lib/python3.7/site-packages/asteval-0.9.13-py3.7.egg/asteval/asteval.py", line 261, in raise_exception
TypeError: at expr='<_ast.Module object at 0x1a263afb00>'
```

**Comment by phlpphns** [comment]
I hope this is the right place to post the issue. I post this as a follow-up. I encounter an error with loading the model if there is a ConstantModel (or custom model?) included.


```python
import numpy as np
import matplotlib.pyplot as plt
from lmfit.model import save_modelresult, load_modelresult
from lmfit.models import VoigtModel, ConstantModel, LinearModel, Model
from lmfit.lineshapes import gaussian, lorentzian
```

This works:

```python
x = np.linspace(0, 20, 501)
y = gaussian(x, 1.1, 8.5, 2) + lorentzian(x, 1.7, 8.5, 1.5)
y = y + np.random.normal(size=len(x), scale=0.025)

model = LinearModel() + \
    VoigtModel(prefix='V1_') + VoigtModel(prefix='V2_')
#pars = model.guess(y, x=x)
pars = model.make_params()
result = model.fit(y, pars, x=x)
save_modelresult(result, 'voigt_modelresult')

print("saved")

r2 = load_modelresult('voigt_modelresult')

print(r2.result.params)
```

This crashes:

```python
x = np.linspace(0, 20, 501)
y = gaussian(x, 1.1, 8.5, 2) + lorentzian(x, 1.7, 8.5, 1.5)
y = y + np.random.normal(size=len(x), scale=0.025)

model = ConstantModel() + \
    VoigtModel(prefix='V1_') + VoigtModel(prefix='V2_')
#pars = model.guess(y, x=x)
pars = model.make_params()
result = model.fit(y, pars, x=x)
save_modelresult(result, 'voigt_modelresult')

print("saved")

r2 = load_modelresult('voigt_modelresult')

print(r2.result.params)
```

This crashes as well:

```python
def func_my_const(x):
    return 0

x = np.linspace(0, 20, 501)
y = gaussian(x, 1.1, 8.5, 2) + lorentzian(x, 1.7, 8.5, 1.5)
y = y + np.random.normal(size=len(x), scale=0.025)

model = Model(func_my_const, prefix="my_const") + \
    VoigtModel(prefix='V1_') + VoigtModel(prefix='V2_')
#pars = model.guess(y, x=x)
pars = model.make_params()
result = model.fit(y, pars, x=x)
save_modelresult(result, 'voigt_modelresult')

print("saved")

r2 = load_modelresult('voigt_modelresult')

print(r2.result.params)
```


Giving an error:

```
Traceback (most recent call last):
  File "/home/philipp/amubox_nextcloud/00_orga/publication_epcos/script_load_and_process_lmfit_results.py", line 340, in <module>
    r2 = load_modelresult('voigt_modelresult')
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1282, in load_modelresult
    mresult = modres.load(fh, funcdefs=funcdefs)
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1769, in load
    return self.loads(fp.read(), funcdefs=funcdefs, **kws)
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1703, in loads
    self.model = _buildmodel(decode4js(modres['model']), funcdefs=funcdefs)
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1243, in _buildmodel
    lmodel = _buildmodel(left, funcdefs=funcdefs)
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1243, in _buildmodel
    lmodel = _buildmodel(left, funcdefs=funcdefs)
  File "/home/philipp/miniconda3/lib/python3.8/site-packages/lmfit/model.py", line 1225, in _buildmodel
    raise ValueError("Cannot restore Model: model function not found")
ValueError: Cannot restore Model: model function not found
```

**Comment by reneeotten** [comment]
> I hope this is the right place to post the issue. 

not really, hijacking an issue that was closed about 2.5 years ago doesn't sound like the right place. In fact, we clearly say everywhere to start a discussion on the mainlinglist or Discussions page on GitHub.

**Comment by newville** [comment]
@phlpphns @reneeotten I can believe that report indicates a different problem. It would be helpful to see a report as a new issue, and including the traceback.   I suspect that you need to include the function definition when loading the modelresult, for example, 

```
def my_constant(x, c=0):
    return c

model = Model(my_constant) + VoigtModel()
params = model.make_params(....)
result = model.fit(y, params, x=x)
save_modelresult(result, 'modelresult.txt') 

restored_result = load_modelresult('modelresult.txt', funcdefs={'my_constant': my_constant})
```

**Comment by phlpphns** [comment]
Thank you, for your answers.  @reneeotten's one made me think and in the meanwhile I extracted from it that my post was not appreciated. Followingly, I deleted it.

Thank you very much for your answer, Matt @newville . While I found a workaround via pickling, I updated lmfit via conda after my "bugreport" on Saturday to the back then two days old version 1.0.3. I freely admit that the other way round would have been better. For the new configuration, 

lmfit: 1.0.3, scipy: 1.6.2, numpy: 1.19.2,asteval: 0.9.23, uncertainties: 3.1.5

, my two erroneous cases and also the code proposed by Matt work through, so I guess I should abstain from opening an issue.

If I would find another issue in the future, I will come back by opening the "appropriate" kind of thread, such as desired by you. Thanks to Matt's answer, I will use lmfit as happily as before and share if I have to contribute something.
